cobalt/browser/metrics/cobalt_metrics_logs_uploader.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright 2025 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/browser/metrics/cobalt_metrics_logs_uploader.h"
+#include "base/base64url.h"
+#include "base/logging.h"
+#include "cobalt/browser/h5vcc_metrics/public/mojom/h5vcc_metrics.mojom.h"
+#include "components/metrics/log_decoder.h"
+#include "components/metrics/metrics_log_uploader.h"
+#include "mojo/public/cpp/bindings/remote.h"
+#include "third_party/metrics_proto/chrome_user_metrics_extension.pb.h"
+#include "third_party/metrics_proto/cobalt_uma_event.pb.h"
+#include "third_party/metrics_proto/reporting_info.pb.h"
+
+namespace cobalt {
+
+using browser::metrics::CobaltUMAEvent;
+
+// Helper method to create a trimmed down and sanitized version of the UMA
+// proto for Cobalt
+void PopulateCobaltUmaEvent(
+    const ::metrics::ChromeUserMetricsExtension& uma_proto,
+    const ::metrics::ReportingInfo reporting_info_proto,
+    CobaltUMAEvent& cobalt_proto) {
+  cobalt_proto.mutable_histogram_event()->CopyFrom(uma_proto.histogram_event());
+  cobalt_proto.mutable_user_action_event()->CopyFrom(
+      uma_proto.user_action_event());
+  cobalt_proto.mutable_reporting_info()->CopyFrom(reporting_info_proto);
+}
+
+void CobaltMetricsLogUploader::UploadLog(
+    const std::string& compressed_log_data,
+    const std::string& log_hash,
+    const std::string& log_signature,
+    const metrics::ReportingInfo& reporting_info) {
+  if (metrics_listener_.is_bound()) {
+    if (service_type_ == ::metrics::MetricsLogUploader::UMA) {
```

Comment:
Ideally we'd use return early here, i.e. 
```
  if (!metrics_listner_.is_bound()){
    return;
  }
  if (service_type_ != ...) {
```
but you want to always run the `on_upload_complete_`./ A trick for that
is `base::ScopedClosureRunner`:
```
  base::ScopedClosureRunner closure(base::BindOnce(on_upload_complete_, ...arguments...));
```
that will automatically run regardless of the return path.